File: typescript/mars-rover/src/mars-rover.ts
interface IPosition {
    x: number;
    y: number;
}

export enum Bearing {
    NORTH,
    EAST,
    SOUTH,
    WEST,
}

export class MarsRover {
    constructor(readonly bearing: Bearing = Bearing.NORTH,
                readonly position: IPosition = {x: 0, y: 0}) {
    }

    private static nextBearing(bearing: Bearing) {
        switch (bearing) {
            case Bearing.NORTH:
                return Bearing.EAST;
            case Bearing.EAST:
                return Bearing.SOUTH;
            case Bearing.SOUTH:
                return Bearing.WEST;
            case Bearing.WEST:
                return Bearing.NORTH;
        }
    }

    private static nextPosition(bearing: Bearing, position: IPosition) {
        switch (bearing) {
            case Bearing.NORTH:
                return MarsRover.deltaY(position, -1);
            case Bearing.SOUTH:
                return MarsRover.deltaY(position, 1);
            case Bearing.EAST:
                return MarsRover.deltaX(position, 1);
            case Bearing.WEST:
                return MarsRover.deltaX(position, -1);
        }
    }

    private static deltaX(position: IPosition, delta: number) {
        return {x: position.x + delta, y: position.y};
    }

    private static deltaY(position: IPosition, delta: number) {
        return {x: position.x, y: position.y + delta};
    }

    forward() {
        return new MarsRover(this.bearing, MarsRover.nextPosition(this.bearing, this.position));
    }

    turnRight() {
        return new MarsRover(MarsRover.nextBearing(this.bearing), this.position);
    }

    backward() {
        return this.turnRight().turnRight().forward().turnRight().turnRight();
    }

    turnLeft() {
        return this.turnRight().turnRight().turnRight();
    }

    move(commandList: string) {
        return commandList
            .split('')
            .map(c => this.translate(c))
            .reduce((acc, command) => command.call(acc), this);
    }

    private translate(command: any) {
        switch (command) {
            case 'f':
                return this.forward;
            case 'b':
                return this.backward;
            case 'r':
                return this.turnRight;
            case 'l':
                return this.turnLeft;
        }
    }
}

File: haskell/mars-rover/src/MarsRover.hs
{-# LANGUAGE RecordWildCards #-}

module MarsRover
  ( mkRover
  , commands
  , position
  , facing
  , Facing(..)
  , Rover
  ) where

import           Control.Monad (foldM)
import           Data.Maybe    (fromMaybe)

data Rover = Rover
  { position :: Position
  , facing   :: Facing
  } deriving (Show, Eq)

data Facing
  = North
  | West
  | South
  | East
  deriving (Show, Eq)

type Position = (Int, Int)

mkRover :: Rover
mkRover = Rover {position = (0, 0), facing = North}

commands :: Rover -> String -> Rover
commands rover = fromMaybe rover . foldM (flip command) rover

command :: Char -> Rover -> Maybe Rover
command 'f' = Just . forward
command 'l' = Just . turnLeft
command 'r' = Just . turnRight
command 'b' = Just . backward
command _   = const Nothing

forward :: Rover -> Rover
forward rover@Rover {..} = rover {position = newPosition position facing}
  where
    newPosition (x, y) North = (x, y + 1)
    newPosition (x, y) South = (x, y - 1)
    newPosition (x, y) West  = (x + 1, y)
    newPosition (x, y) East  = (x - 1, y)

turnLeft :: Rover -> Rover
turnLeft rover@Rover {..} = rover {facing = newFacing facing}
  where
    newFacing North = West
    newFacing West  = South
    newFacing South = East
    newFacing East  = North

turnRight :: Rover -> Rover
turnRight = turnLeft . turnLeft . turnLeft

backward :: Rover -> Rover
backward = turnLeft . turnLeft . forward . turnLeft . turnLeft
